### PR TITLE
NativeAOT-LLVM: Merge to August 2022

### PR DESCRIPTION
**Pull request by yowl**
Contrary to the branch name this PR merges up to August, commit 442c137891821a567e9a05411f821dbf2aec5aa5

LoongArch addition, some test changes.

partial implementation of `GT_STORE_LCL_FLD` - there are some TODO's left .  The merge is large already so can leave the TODO's for later or do them here?

Excluded some of the new tests that failed.

Removed the Linux release job as it was failing - maybe more merging will make it work again...

### CODE REVIEW

File: docs/workflow/building/coreclr/nativeaot.md [hunk 1]
```diff
@@ -37,7 +37,6 @@ The executable looks like a native executable, in the sense that it can be debug
 
 The compiler also has a mode where each managed assembly can be compiled into a separate object file. The object files are later linked into a single executable using the platform linker. This mode is mostly used in testing (it's faster to compile this way because we don't need to recompiling the same code from e.g. CoreLib). It's not a shipping configuration and has many problems (requires exactly matching compilation settings, forfeits many optimizations, and has trouble around cross-module generic virtual method implementations).
 
-<<<<<<< HEAD
 ## Building
```
Comment (SingleAccretion) [inline docs/workflow/building/coreclr/nativeaot.md]: We should delete this section, it was superseded by `Building` at the top of the document:
```diff
-## Building
 ## Building for Web Assembly
```

File: src/coreclr/jit/compiler.cpp [hunk 2]
```diff
@@ -4755,11 +4719,13 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
     fgDebugCheckLinks();
 #endif
 
+#ifndef TARGET_WASM
     // Morph multi-dimensional array operations.
     // (Consider deferring all array operation morphing, including single-dimensional array ops,
     // from global morph to here, so cloning doesn't have to deal with morphed forms.)
     //
     DoPhase(this, PHASE_MORPH_MDARR, &Compiler::fgMorphArrayOps);
```
Comment (SingleAccretion) [inline src/coreclr/jit/compiler.cpp:4726]: Why is this phase under `!TARGET_WASM`? It should not be; it is the way we get MD array support "for free".

File: src/coreclr/jit/compiler.h [hunk 3]
```diff
@@ -4579,9 +4576,7 @@ class Compiler
 
     bool fgMorphBlockStmt(BasicBlock* block, Statement* stmt DEBUGARG(const char* msg));
 
-<<<<<<< HEAD
 #ifndef TARGET_WASM
-=======
     //------------------------------------------------------------------------------------------------------------
     // MorphMDArrayTempCache: a simple cache of compiler temporaries in the local variable table, used to minimize
     // the number of locals allocated when doing early multi-dimensional array operation expansion. Two types of
```
Comment (SingleAccretion) [inline src/coreclr/jit/compiler.h]: Ditto: MD arrays.

File: src/coreclr/jit/gentree.cpp [hunk 4]
```diff
@@ -2964,8 +2964,6 @@ unsigned Compiler::gtHashValue(GenTree* tree)
                         unreached();
                     }
                 }
-
-                add = genTreeHashAdd(ulo32(add), vecCon->GetSimdBaseType());
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp]: Was this deletion intentional?

File: src/coreclr/jit/llvm.cpp [hunk 5]
```diff
@@ -395,6 +393,10 @@ bool Llvm::helperCallHasManagedCallingConvention(CorInfoHelpFunc helperFunc) con
         { FUNC(CORINFO_HELP_CHKCASTCLASS) CORINFO_TYPE_CLASS, { CORINFO_TYPE_PTR, CORINFO_TYPE_CLASS } },
         { FUNC(CORINFO_HELP_CHKCASTANY) CORINFO_TYPE_CLASS, { CORINFO_TYPE_PTR, CORINFO_TYPE_CLASS } },
         { FUNC(CORINFO_HELP_CHKCASTCLASS_SPECIAL) CORINFO_TYPE_CLASS, { CORINFO_TYPE_PTR, CORINFO_TYPE_CLASS } },
+
+        // TODO-LLVM: jithelpers.cpp, this is a bit of a guess...
+        { FUNC(CORINFO_HELP_ISINSTANCEOF_EXCEPTION) CORINFO_TYPE_BOOL, { CORINFO_TYPE_RT_HANDLE, CORINFO_TYPE_CLASS} },
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]:
```suggestion
        { FUNC(CORINFO_HELP_ISINSTANCEOF_EXCEPTION) CORINFO_TYPE_BOOL, { CORINFO_TYPE_PTR, CORINFO_TYPE_CLASS } },
```
Implementation: https://github.com/dotnet/runtime/blob/2ee941b5e33b68fc39d2c6c2a26f025e0d2efee6/src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs#L932.

File: src/coreclr/jit/llvm.cpp [hunk 6]
```diff
@@ -592,6 +594,9 @@ bool Llvm::helperCallHasManagedCallingConvention(CorInfoHelpFunc helperFunc) con
         // Dead code.
         { FUNC(CORINFO_HELP_THROW_TYPE_NOT_SUPPORTED) },
 
+        // TODO-LLVM:
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]:
```suggestion
        // Not used in NativeAOT.
```
Source: https://github.com/dotnet/runtime/blob/2ee941b5e33b68fc39d2c6c2a26f025e0d2efee6/src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs#L452.

File: src/coreclr/jit/llvmcodegen.cpp [hunk 7]
```diff
@@ -1393,42 +1393,42 @@ void Llvm::buildEmptyPhi(GenTreePhi* phi)
...
 {
     GenTree* data = lclFld->gtGetOp1();
-    Type* llvmStoreType = getLlvmTypeForVarType(lclFld->TypeGet());
-    // TODO-LLVM: uncomment once we merge https://github.com/dotnet/runtime/pull/68986 (May 27).
-    // ClassLayout* layout = lclFld->TypeIs(TYP_STRUCT) ? lclFld->GetLayout() : nullptr;
-    // Type* llvmStoreType = (layout != nullptr) ? getLlvmTypeForStruct(lclFld->GetLayout())
-    //                                           : getLlvmTypeForVarType(lclFld->TypeGet());
+
+    ClassLayout* layout = lclFld->TypeIs(TYP_STRUCT) ? lclFld->GetLayout() : nullptr;
+    Type* llvmStoreType = (layout != nullptr) ? getLlvmTypeForStruct(lclFld->GetLayout())
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmcodegen.cpp]:
```suggestion
    Type* llvmStoreType = (layout != nullptr) ? getLlvmTypeForStruct(layout)
```

File: src/coreclr/jit/llvmcodegen.cpp [hunk 8]
```diff
@@ -1393,42 +1393,42 @@ void Llvm::buildEmptyPhi(GenTreePhi* phi)
...
 
 void Llvm::buildStoreLocalField(GenTreeLclFld* lclFld)
 {
     GenTree* data = lclFld->gtGetOp1();
-    Type* llvmStoreType = getLlvmTypeForVarType(lclFld->TypeGet());
-    // TODO-LLVM: uncomment once we merge https://github.com/dotnet/runtime/pull/68986 (May 27).
-    // ClassLayout* layout = lclFld->TypeIs(TYP_STRUCT) ? lclFld->GetLayout() : nullptr;
-    // Type* llvmStoreType = (layout != nullptr) ? getLlvmTypeForStruct(lclFld->GetLayout())
-    //                                           : getLlvmTypeForVarType(lclFld->TypeGet());
+
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmcodegen.cpp]:
```suggestion
```

File: src/coreclr/jit/llvmcodegen.cpp [hunk 9]
```diff
@@ -1393,42 +1393,42 @@ void Llvm::buildEmptyPhi(GenTreePhi* phi)
...
-        // {
-        //     assert(data->OperIsInitVal());
-        //     Value* fillValue = consumeValue(data->gtGetOp1(), Type::getInt8Ty(_llvmContext));
-        //     Value* sizeValue = _builder.getInt32(layout->GetSize());
-        //     _builder.CreateMemSet(addrValue, fillValue, sizeValue, llvm::MaybeAlign());
-        //     return;
-        // }
-        //
-        // dataValue = llvm::Constant::getNullValue(llvmStoreType);
+        ClassLayout* layout = lclFld->GetLayout();
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmcodegen.cpp]:
```suggestion
```

File: src/coreclr/jit/llvmcodegen.cpp [hunk 10]
```diff
@@ -1393,42 +1393,42 @@ void Llvm::buildEmptyPhi(GenTreePhi* phi)
...
+        if (!data->IsIntegralConst(0))
+        {
+            assert(data->OperIsInitVal());
+            Value* fillValue = consumeValue(data->gtGetOp1(), Type::getInt8Ty(_llvmContext));
+            Value* sizeValue = _builder.getInt32(layout->GetSize());
+            _builder.CreateMemSet(addrValue, fillValue, sizeValue, llvm::MaybeAlign());
+            return;
+        }
+
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmcodegen.cpp]:
```suggestion

```

File: src/coreclr/jit/llvmlower.cpp [hunk 11]
```diff
@@ -475,10 +475,10 @@ void Llvm::lowerStoreLcl(GenTreeLclVarCommon* storeLclNode)
     LclVarDsc* varDsc = _compiler->lvaGetDesc(lclNum);
     GenTree* data = storeLclNode->gtGetOp1();
 
-    unsigned convertToStoreObjLclNum = BAD_VAR_NUM;
+    unsigned convertToStoreLclFldVarNum = BAD_VAR_NUM;
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmlower.cpp]: Nit: `convertToStoreLclFldVarNum` -> `convertToStoreLclFldLclNum` (`lclNum` is more prevalent than `varNum`).

File: src/coreclr/jit/llvmlower.cpp [hunk 12]
```diff
@@ -489,22 +489,25 @@ void Llvm::lowerStoreLcl(GenTreeLclVarCommon* storeLclNode)
...
+        storeLclNode->ChangeType(lclFldVarType);
+        storeLclNode->SetLclNum(convertToStoreLclFldVarNum);
+        GenTreeLclFld* storeLclFldNode = storeLclNode->AsLclFld();
+        storeLclFldNode->SetLclOffs(lclFldVarDsc->lvFldOffset);
+        if (varTypeIsStruct(lclFldVarType))
+        {
+            storeLclFldNode->SetLayout(lclFldVarDsc->GetLayout());
+        }
+        storeLclFldNode->gtFlags |= (GTF_ASG | GTF_IND_NONFAULTING);
+        storeLclFldNode->gtOp1 = data;
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmlower.cpp]:
```suggestion
        storeLclNode->SetOper(GT_STORE_LCL_FLD);
        storeLclNode->SetLclNum(convertToStoreLclFldVarNum);
        storeLclNode->AsLclFld()->SetLclOffs(0);
        storeLclNode->AsLclFld()->SetLayout(varDsc->GetLayout());
```

File: src/coreclr/jit/morph.cpp [hunk 13]
```diff
@@ -17218,6 +17169,7 @@ GenTree* Compiler::fgMorphReduceAddOps(GenTree* tree)
     return morphed;
 }
 
+#ifndef TARGET_WASM
 //------------------------------------------------------------------------
 // Compiler::MorphMDArrayTempCache::TempList::GetTemp: return a local variable number to use as a temporary variable
```
Comment (SingleAccretion) [inline src/coreclr/jit/morph.cpp:17158]: Ditto: MD arrays.

File: src/coreclr/nativeaot/BuildIntegration/BuildFrameworkNativeObjects.proj [hunk 14]
```diff
@@ -10,11 +10,8 @@
     <DebugSymbols Condition="'$(DebugSymbols)' == ''">true</DebugSymbols>
   </PropertyGroup>
 
-<<<<<<< HEAD
   <Import Project="$(MSBuildThisFileDirectory)\Microsoft.DotNet.ILCompiler.LLVM.targets" Condition="'$(IlcCalledViaPackage)' == 'true'"/>
-=======
   <Import Project="$(MSBuildThisFileDirectory)\Microsoft.DotNet.ILCompiler.SingleEntry.targets" Condition="'$(IlcCalledViaPackage)' == 'true'" />
->>>>>>> 442c137891821a567e9a05411f821dbf2aec5aa5
```
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/BuildFrameworkNativeObjects.proj]: Hmm, it does not make sense to import both of these. Presumably, only the `.SingleEntry` one should be imported.

Additionally, `Microsoft.DotNet.ILCompiler.LLVM.targets` is now importing `Microsoft.DotNet.ILCompiler.targets`, which does not exist.

File: src/coreclr/nativeaot/Runtime/thread.cpp [hunk 15]
```diff
@@ -604,6 +604,7 @@ void Thread::Hijack()
     PalHijack(m_hPalThread, this);
 }
 
+#if !defined(HOST_WASM) // Wasm cannot get the instruction pointer
```
Comment (SingleAccretion) [inline src/coreclr/nativeaot/Runtime/thread.cpp]:
```suggestion
#if !defined(HOST_WASM)
```
Hijacking in general does not make sense on WASM.

<details>
<summary>What is "hijacking" in this context?</summary>

On native architectures, it is possible to suspend running managed code by overwriting the return address stored on the stack with a pointer to the EE's "wait for GC" routine - "hijacking" the thread.
</details>

File: src/coreclr/nativeaot/Runtime/thread.cpp [hunk 16]
```diff
@@ -750,6 +752,7 @@ void Thread::HijackReturnAddress(NATIVE_CONTEXT* pSuspendCtx, HijackFunc* pfnHij
     HijackReturnAddressWorker(&frameIterator, pfnHijackFunction);
 }
 
+#if !defined(HOST_WASM) // Wasm has no implementation for ReturnKindToTransitionFrameFlags
```
Comment (SingleAccretion) [inline src/coreclr/nativeaot/Runtime/thread.cpp]:
```suggestion
#if !defined(HOST_WASM)
```

File: src/coreclr/nativeaot/Runtime/threadstore.cpp [hunk 17]
```diff
@@ -82,8 +82,11 @@ ThreadStore * ThreadStore::Create(RuntimeInstance * pRuntimeInstance)
     if (NULL == pNewThreadStore)
         return NULL;
 
+    // TODO-LLVM: will wasm threads help with this?
```
Comment (SingleAccretion) [inline src/coreclr/nativeaot/Runtime/threadstore.cpp]:
```suggestion
```
(I doubt WASM will ever be capable of using this suspension technique)

File: src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp [hunk 18]
```diff
@@ -997,7 +997,12 @@ REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalRegisterHijackCallback(_In_ PalH
 
 REDHAWK_PALEXPORT void REDHAWK_PALAPI PalHijack(HANDLE hThread, _In_opt_ void* pThreadToHijack)
 {
-    ThreadUnixHandle* threadHandle = (ThreadUnixHandle*)hThread;
+// TODO-LLVM: enable if we implement wasm threads
+#if HOST_WASM
+    abort();
+#endif
+
```
Comment (SingleAccretion) [inline src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp]: Do we need this abort or can we leave this code alone to be deleted by the compiler?
```suggestion
```

File: src/tests/nativeaot/SmokeTests/UnitTests/Interfaces.cs [hunk 19]
```diff
@@ -40,33 +40,30 @@ public static int Run()
...
-=======
         TestSimpleStaticDefaultInterfaceMethods.Run();
+#if !CODEGEN_WASM // TODO-LLVM: fails because the scanner for LLVM adds a ScannedMethodNode node on the static abstract interface method (which asserts)
         TestSimpleDynamicStaticVirtualMethods.Run();
         TestGenericDynamicStaticVirtualMethods.Run();
         TestVariantGenericDynamicStaticVirtualMethods.Run();
         TestStaticDefaultMethodAmbiguity.Run();
+#endif
         TestMoreConstraints.Run();
+#if !CODEGEN_WASM // TODO-LLVM: fails because the scanner for LLVM adds a ScannedMethodNode node on the static abstract interface method (which asserts)
```
Comment (SingleAccretion) [inline src/tests/nativeaot/SmokeTests/UnitTests/Interfaces.cs]: Can't wait to re-enable all these...

**Review by SingleAccretion — COMMENTED** [review]
There are a number of great things in this merge for us; thank you for working on it!

A couple suggestions / questions, one concern about the MD array code.

**Comment by SingleAccretion** [comment]
>  leave the TODO's for later?

Seems best.

File: src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp (hunk 18 above)
Comment (yowl) [reply src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp]: We can remove it, thanks

File: src/coreclr/jit/llvmlower.cpp (hunk 12 above)
Comment (yowl) [reply src/coreclr/jit/llvmlower.cpp]: thats simpler!  thanks.

File: src/coreclr/nativeaot/Runtime/threadstore.cpp (hunk 17 above)
Comment (yowl) [reply src/coreclr/nativeaot/Runtime/threadstore.cpp]: removed.

File: src/coreclr/nativeaot/Runtime/thread.cpp (hunk 16 above)
Comment (yowl) [reply src/coreclr/nativeaot/Runtime/thread.cpp]: Removed

File: src/coreclr/nativeaot/Runtime/thread.cpp (hunk 15 above)
Comment (yowl) [reply src/coreclr/nativeaot/Runtime/thread.cpp]: Removed.

File: src/coreclr/nativeaot/BuildIntegration/BuildFrameworkNativeObjects.proj (hunk 14 above)
Comment (yowl) [reply src/coreclr/nativeaot/BuildIntegration/BuildFrameworkNativeObjects.proj]: Publishing still looks broken:
```
E:\tmp\consolelocal\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-dev\build\Microsoft.NETCore.Native.Publish.targets
(78,5): error : The PrivateSdkAssemblies ItemGroup is required for _ComputeAssembliesToCompileToNative [E:\tmp\consolel
ocal\console.csproj]
```
Think we had this problem with the tests in the last merge.

File: src/coreclr/jit/llvmlower.cpp (hunk 11 above)
Comment (yowl) [reply src/coreclr/jit/llvmlower.cpp]: done

File: src/coreclr/jit/gentree.cpp (hunk 4 above)
Comment (yowl) [reply src/coreclr/jit/gentree.cpp]: no, reinstated and the whole case excluded for `TARGET_WASM` as the flow through `unreached` causes an unreachable code error

File: src/coreclr/jit/compiler.cpp (hunk 2 above)
Comment (yowl) [reply src/coreclr/jit/compiler.cpp:4726]: Reinstated, thanks.

File: src/coreclr/jit/compiler.h (hunk 3 above)
Comment (yowl) [reply src/coreclr/jit/compiler.h]: Reinstated

File: src/coreclr/jit/morph.cpp (hunk 13 above)
Comment (yowl) [reply src/coreclr/jit/morph.cpp:17158]: Reinstated

File: src/coreclr/nativeaot/BuildIntegration/BuildFrameworkNativeObjects.proj (hunk 14 above)
Comment (SingleAccretion) [reply src/coreclr/nativeaot/BuildIntegration/BuildFrameworkNativeObjects.proj]: Yes, we'll need to fix publish more thoroughly after this merge (which changes a few things). Does it break the tests if we remove the "nop" import here?

Comment (yowl) [reply src/coreclr/nativeaot/BuildIntegration/BuildFrameworkNativeObjects.proj]: No, it seems fine.  Removed `Microsoft.DotNet.ILCompiler.SingleEntry.targets`

**Review by SingleAccretion — COMMENTED** [review]
Code changes LGTM; looks like tests found an AV though.

**Comment by yowl** [comment]
> Code changes LGTM; looks like tests found an AV though.

Yes, there's a couple of things I think.  The AV goes if I put back the code in `lowerStoreLcl`.  I think the other version might be passing null to `SetLayout` sometimes.  Not looked in detail at that yet.

```cs
  storeLclNode->SetOper(GT_STORE_LCL_FLD);
  LclVarDsc* lclFldVarDsc  = _compiler->lvaGetDesc(convertToStoreLclFldLclNum);
  var_types lclFldVarType = lclFldVarDsc->TypeGet();
  storeLclNode->ChangeType(lclFldVarType);
  storeLclNode->SetLclNum(convertToStoreLclFldLclNum);
  GenTreeLclFld* storeLclFldNode = storeLclNode->AsLclFld();
  storeLclFldNode->SetLclOffs(lclFldVarDsc->lvFldOffset);
  if (varTypeIsStruct(lclFldVarType))
  {
      storeLclFldNode->SetLayout(lclFldVarDsc->GetLayout());
  }
  storeLclFldNode->gtFlags |= (GTF_ASG | GTF_IND_NONFAULTING);
  storeLclFldNode->gtOp1 = data;
```
~~I may also have dropped `GetNativeFunctionToCall` somewhere, but not sure on that, still testing~~

**Review by SingleAccretion — APPROVED** [review]
🚀

**Comment by yowl** [comment]
@SingleAccretion  you were ok with what `lowerStoreLcl` ended up with ?

**Comment by SingleAccretion** [comment]
Yes. It is a little strange, but let's get this change in and I will follow up on it locally.

**Comment by yowl** [comment]
ok, thanks for the review.

**Review by jkotas — COMMENTED** [review]
Nice!

**Pull request merged**